hyperspy/drawing/markers.py

```diff
@@ -0,0 +1,816 @@
+        ...    collection=PolyCollection,
+        ...    offsets=np.random.rand(10,2)*10,
+        ...    verts=np.array([[0,0], [0,1], [1,1], [1,0]]),
+        ...    color="red",
+        ...    )
+        >>>s = hs.signals.Signal2D(np.ones((10,10,10,10)))
+        >>>s.plot()
+        >>>s.add_marker(m)
+
+        """
+        if isinstance(collection, str):
+            try:
+                collection = getattr(mpl_collections, collection)
+            except AttributeError:
+                raise ValueError(
+                    f"'{collection}' is not the name of a matplotlib collection class."
+                 )
+
+        if not issubclass(collection, mpl_collections.Collection):
+            raise ValueError(
+                f"{collection} is not a subclass of `matplotlib.collection.Collection`."
+                )
+
+        if ".".join(collection.__module__.split('.')[:2]) not in [
+            'matplotlib.collections', 'hyperspy.external'
+            ]:
+            # To be able to load a custom markers, we need to be able to instantiate
+            # the class and the safe way to do that is to import from
+            # `matplotlib.collections` or `hyperspy.external` (patched matplotlib collection)
+            raise ValueError(
+                "To support loading file saved with custom markers, the collection must be "
+                "implemented in matplotlib or hyperspy"
+            )
+
+        # Data attributes
+        self.kwargs = kwargs  # all keyword arguments.
+        self.axes_manager = None
+        self.ax = None
+        self.auto_update = True
+        self._offset_transform = None
+        self._transform = None
+
+        # Handling dask arrays
+        self.dask_kwargs = {}
+        for key, value in self.kwargs.items():
+            if isinstance(value, da.Array) and value.dtype == object:
+                self.dask_kwargs[key] = self.kwargs[key]
+            elif isinstance(value, da.Array):  # and value.dtype != object:
+                self.kwargs[key] = value.compute()
+            # Patches or verts shouldn't be cast to array
+            elif (
+                isinstance(value, list)
+                and len(value) > 0
+                and not isinstance(value[0], Patch)
+                and not key == "verts"
+            ):
+                self.kwargs[key] = np.array(value)
+            elif isinstance(value, list) and len(value) == 0:
+                self.kwargs[key] = np.array(value)
+
+            if key in ["sizes", "color"] and (
+                not hasattr(value, "__len__") or isinstance(value, str)
+            ):
+                self.kwargs[key] = (value,)
+            if key in ["s"] and isinstance(value, str):
+                self.kwargs[key] = (value,)
+
+        self._cache_dask_chunk_kwargs = {}
+        self._cache_dask_chunk_kwargs_slice = {}
+
+        self._class_name = self.__class__.__name__
+        self.name = name
+        # Properties
+        self.collection = None
```

Comment:
I am wondering if we should privatise the `collection` attribute? For example, if a user changes a transform through the matplotlib collection directly, save the signal and try to reload it, it will not reload in the same stage as it was saved, because the transform needs to be set through the corresponding setter in the `Markers` class for this to work.

Reply:
That seems reasonable.  I think that we have given enough flexibility for most people to be happy with the current state.  I think that if people want more ability to change their plots we should probably just try to open up the plotting API (In a different PR and after the 2.0.0 release).